euphoria-local/src/main/java/cz/seznam/euphoria/executor/local/LocalExecutor.java

```diff
@@ -719,17 +717,19 @@ private InputProvider execReduceStateByKey(
               final Iterable<Window> targetWindows;
               int windowShift = 0;
               if (allowWindowBasedShuffling) {
-                if (windowing.isPresent()) {
+                if (windowing != null) {
                   // FIXME: the time function inside windowing
                   // must be part of the operator itself
-                  targetWindows = windowing.get().assignWindowsToElement(datum);
+                  targetWindows = windowing.assignWindowsToElement(datum);
                 } else {
                   targetWindows = Collections.singleton(datum.getWindow());
                 }
 
                 if (!isMergingWindowing && Iterables.size(targetWindows) == 1) {
-                  windowShift = new Random(
-                      Iterables.getOnlyElement(targetWindows).hashCode()).nextInt();
+                  final int windowHash = Iterables.getOnlyElement(targetWindows).hashCode();
+                  synchronized (random) {
+                    windowShift = shifts.computeIfAbsent(windowHash, (ignore) -> random.nextInt());
+                  }
```

Comment:
I'm not sure, that we are trading time for space at reasonable exchange ratio. :)
The original code might be a little more CPU ang GC intensive, while this code will have to hold possible unbounded map in memory. I vote for the previous implementation.

Reply:
How about setting an upper bound for this? `windowShift = shifts.computeIfAbsent(windowHash % BOUND, (ignore) -> random.nextInt());`